lib/policies/timestamp-generation.js

```diff
@@ -0,0 +1,148 @@
+ * @param {Client} client The {@link Client} instance to generate timestamps to.
+ * @returns {Long|Number|null} the next timestamp (in microseconds). If it's equals to <code>null</code>, it won't be
+ * sent by the driver, letting the server to generate the timestamp.
+ * @abstract
+ */
+TimestampGenerator.prototype.next = function (client) {
+  throw new Error('next() must be implemented');
+};
+
+/**
+ * A timestamp generator that guarantees monotonically increasing timestamps and logs warnings when timestamps
+ * drift in the future.
+ * <p>
+ *   {@link Date} has millisecond precision and client timestamps require microsecond precision. This generator
+ *   keeps track of the last generated timestamp, and if the current time is within the same millisecond as the last,
+ *   it fills the microsecond portion of the new timestamp with the value of an incrementing counter.
+ * </p>
+ * @param {Number} [warningThreshold] Determines how far in the future timestamps are allowed to drift before a
+ * warning is logged, expressed in milliseconds. Default: <code>1000</code>.
+ * @param {Number} [minLogInterval] In case of multiple log events, it determines the time separation between log events,
+ * expressed in milliseconds. Use 0 to disable. Default: <code>1000</code>.
+ * @extends {TimestampGenerator}
+ * @constructor
+ */
+function MonotonicTimestampGenerator(warningThreshold, minLogInterval) {
+  if (warningThreshold < 0) {
+    throw new errors.ArgumentError('warningThreshold can not be lower than 0');
+  }
+  this._warningThreshold = warningThreshold || 1000;
+  this._minLogInterval = minLogInterval || 1000;
+  this._micros = -1;
+  this._lastDate = 0;
+  this._lastLogDate = 0;
+}
+
+util.inherits(MonotonicTimestampGenerator, TimestampGenerator);
+
+/**
+ * Returns the current time in milliseconds since UNIX epoch
+ * @returns {Number}
+ */
+MonotonicTimestampGenerator.prototype.getDate = function () {
+  return Date.now();
+};
+
+MonotonicTimestampGenerator.prototype.next = function (client) {
+  var date = this.getDate();
+  var drifted = 0;
+  if (date > this._lastDate) {
+    this._micros = 0;
+    this._lastDate = date;
+    return this._generateMicroseconds();
+  }
+
+  if (date < this._lastDate) {
+    drifted = this._lastDate - date;
+    date = this._lastDate;
+  }
+  if (++this._micros === 1000) {
+    this._micros = 0;
+    if (date === this._lastDate) {
+      // Move date 1 millisecond into the future
+      date++;
+      drifted++;
+    }
+  }
+  if (drifted >= this._warningThreshold) {
+    // Avoid logging an unbounded amount of times within a clock-skew event or during an interval when more than 1
+    // query is being issued by microsecond
+    var currentLogDate = Date.now();
+    if (this._minLogInterval > 0 && this._lastLogDate + this._minLogInterval <= currentLogDate){
+      var message = util.format(
+        'Timestamp generated using current date was %d milliseconds behind the last generated timestamp, returned ' +
```

Comment:
It would be nice to also include the timestamp values (the last generated one and the new one)